src/app/codegen-interaction-model/CodegenDataModel.cpp

```diff
@@ -0,0 +1,411 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <app/codegen-interaction-model/CodegenDataModel.h>
+
+#include <app-common/zap-generated/attribute-type.h>
+#include <app/util/attribute-storage.h>
+#include <app/util/endpoint-config-api.h>
+#include <lib/core/DataModelTypes.h>
+#include <optional>
+
+namespace chip {
+namespace app {
+namespace {
+
+/// Checks if the specified ember cluster mask corresponds to a valid
+/// server cluster.
+bool IsServerMask(EmberAfClusterMask mask)
+{
+    return (mask == 0) || ((mask & CLUSTER_MASK_SERVER) != 0);
```

Comment:
Why are we allowing mask == 0 here?  We shouldn't.

Reply:
It is mock-compatibility it seems:  https://github.com/project-chip/connectedhomeip/blob/master/src/app/util/mock/attribute-storage.cpp#L270 and various other places accept 0 mask even though a real data model should NOT allow this as it seems you are either server or client.

I am afraid that not allowing 0 would lead me to more yak shaving to figure out what to do with mocks.

Reply:
That mock code makes no sense, and importantly is not operating on the mask fields of EmberAfCluster structs.  And it's completely pointless, because all callers of emberAfClusterIndex pass CLUSTER_MASK_SERVER anyway.

Please make this code make sense.   Cluster definitions need to set one of the bits, otherwise everything is broken.

Reply:
Updated - made this a member method and removed the 0 mask check.